_posts/2020-07-12-ray-tune.md

```diff
@@ -0,0 +1,468 @@
+
+[dml-lg]:https://github.com/dask/dask-lightgbm
+[dmlxg]:https://ml.dask.org/xgboost.html
+[Ray's README.md]:https://github.com/ray-project/tune-sklearn/blob/31f228e21ef632a89a74947252d8ad5323cbd043/README.md
+
+### Scale up
+
+*Dask-ML supports distributed tuning (how could it not?). In addition,
+it also supports larger-than-memory data.*
+
+> [Ray's] Tune-sklearn leverages Ray Tune, a library for distributed
+> hyperparameter tuning, to efficiently and transparently parallelize cross
+> validation on multiple cores and even multiple machines.
+
+Naturally, Dask-ML also scales to multiple cores/machines because it relies on
+Dask. Dask has wide support for [different deployment options][ddo] that span
+from your personal machine to supercomputers. Dask will very likely work on top
+of any computing system you have available, including Kubernetes, SLURM, YARN
+and Hadoop clusters as well as your personal machine.
+
+Dask-ML's model selection also scales to larger than memory datasets.
+<!-- TODO talk to Ray devs about this -->
+
+[ray-trainable]:https://medium.com/rapids-ai/30x-faster-hyperparameter-search-with-raytune-and-rapids-403013fbefc5
+
+In addition, I have benchmarked Dask-ML's model selection module to see how the
+time-to-solution is affected by the number of Dask workers in "[Better and
+faster hyperparameter optimization with Dask][db-bf]." That is, how does the
+time to reach a particular accuracy scale with the number of workers $P$? At
+first, it'll scale like $1/P$ but with large number of workers the serial
+portion will dictate time to solution according to [Amdahl's Law]. Briefly, I
+found Dask-ML's `HyperbandSearchCV` speedup started to saturate around 24
+workers for a particular search.
+
+[^bohb-parallel]:In Section 4.2 of [their paper](http://proceedings.mlr.press/v80/falkner18a/falkner18a.pdf).
+
+[db-bf]:https://blog.dask.org/2019/09/30/dask-hyperparam-opt
+
+[Amdahl's Law]:https://en.wikipedia.org/wiki/Amdahl%27s_law
+[ddo]:https://docs.dask.org/en/latest/setup.html
+[plg]:https://lightgbm.readthedocs.io/en/latest/Parallel-Learning-Guide.html
+
+### Speed
+
+*Both Dask-ML and Ray are much faster than Scikit-Learn.*
+
+Ray's tune-sklearn runs some benchmarks in [the introduction][1] with the
+`GridSearchCV` class found Scikit-Learn and Dask-ML. A more fair benchmark
+would be use Dask-ML's `HyperbandSearchCV` because it is almost the same as the
+algorithm in Ray's tune-sklearn. To be specific, I'm interested in comparing
+these methods:
+
+* Scikit-Learn's `RandomizedSearchCV`. This is a popular implementation, one
+  that I've bootstrapped myself with a custom model.
+* Dask-ML's `HyperbandSearchCV`. This is an early stopping technique for
+  `RandomizedSearchCV`.
+* Ray tune-sklearn's `TuneSearchCV`. This is a slightly different early
+  stopping technique than `HyperbandSearchCV`'s.
+
+Each of these will be configured to perform the same tasks, sample 100
```

Comment:
I think this whole section should be in the past tense.

Reply:
Thanks; resolved